File: lumora/app/src/main/java/com/lumora/app/data/AppDatabase.java
package com.lumora.app.data;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

@Database(entities = {TaskEntity.class, HabitEntity.class, ContextLogEntity.class},
        version = 1, exportSchema = false)
public abstract class AppDatabase extends RoomDatabase {

    private static volatile AppDatabase instance;

    public abstract TaskDao taskDao();
    public abstract HabitDao habitDao();
    public abstract ContextLogDao contextLogDao();

    public static AppDatabase get(Context ctx) {
        if (instance == null) {
            synchronized (AppDatabase.class) {
                if (instance == null) {
                    instance = Room.databaseBuilder(ctx.getApplicationContext(),
                                    AppDatabase.class, "lumora.db")
                            .fallbackToDestructiveMigration()
                            .allowMainThreadQueries() // 단순 작업 위주, 호출지점에서 짧게 사용
                            .build();
                }
            }
        }
        return instance;
    }
}

File: EmergencyBlackBox/app/src/main/java/com/emergencyblackbox/AppDatabase.kt
package com.emergencyblackbox

import android.content.Context
import androidx.room.Dao
import androidx.room.Database
import androidx.room.Entity
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.PrimaryKey
import androidx.room.Query
import androidx.room.Room
import androidx.room.RoomDatabase

@Entity(tableName = "recording_events")
data class RecordingEvent(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val filePath: String,
    val startedAtEpochMs: Long,
    val endedAtEpochMs: Long? = null,
    val status: String
)

@Dao
interface RecordingEventDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(event: RecordingEvent): Long

    @Query("UPDATE recording_events SET endedAtEpochMs = :endedAt, status = :status WHERE id = :id")
    suspend fun markCompleted(id: Long, endedAt: Long, status: String)

    @Query("SELECT * FROM recording_events ORDER BY id DESC LIMIT 1")
    suspend fun getLastEvent(): RecordingEvent?
}

@Database(entities = [RecordingEvent::class], version = 1, exportSchema = false)
abstract class AppDatabase : RoomDatabase() {
    abstract fun recordingEventDao(): RecordingEventDao

    companion object {
        @Volatile
        private var instance: AppDatabase? = null

        fun get(context: Context): AppDatabase {
            return instance ?: synchronized(this) {
                instance ?: Room.databaseBuilder(
                    context.applicationContext,
                    AppDatabase::class.java,
                    "emergency_blackbox.db"
                ).build().also { instance = it }
            }
        }
    }
}
